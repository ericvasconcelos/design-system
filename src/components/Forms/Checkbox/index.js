export { CheckboxBasic } from './CheckboxBasic'
export { CheckboxWithFormik } from './CheckboxWithFormik'
