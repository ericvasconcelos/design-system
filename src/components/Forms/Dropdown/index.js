export { DropdownWithFormik } from './DropdownWithFormik'
export { DropdownWithCheckbox } from './DropdownWithCheckbox'
export { DropdownBasic } from './DropdownBasic'
