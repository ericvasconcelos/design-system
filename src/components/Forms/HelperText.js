import React from 'react'

import { Icon } from 'components/Icon'
import { func, oneOf, oneOfType, string } from 'prop-types'
import { isFunction } from 'utils/isFunction'

import { BaseMessage } from './commonStyles'

export const HelperText = ({ status, errorMessage, infoMessage, className }) => {
  if (status === 'error')
    return (
      <BaseMessage className={className} status={status} aria-live="assertive">
        {isFunction(errorMessage) ? errorMessage(errorMessage) : errorMessage}
      </BaseMessage>
    )
  if (infoMessage)
    return (
      <BaseMessage className={className} aria-live="polite">
        {isFunction(infoMessage) ? infoMessage() : (
          <>
            <Icon icon="attentionSmall" size={12} styleColor="darkBlueCash200" />
            {infoMessage}

          </>
        )}
      </BaseMessage>
    )
  return <></>
}

HelperText.propTypes = {
  status: oneOf(['error', 'success', 'default', 'loading']),
  errorMessage: oneOfType([string, func]),
  infoMessage: oneOfType([string, func]),
  className: string
}

HelperText.defaultProps = {
  status: '',
  errorMessage: '',
  infoMessage: '',
  className: ''
}
