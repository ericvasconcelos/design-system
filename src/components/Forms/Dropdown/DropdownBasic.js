import React, { useState, useEffect, useRef, useCallback } from 'react'

import { BaseWrapper } from 'components/Forms/commonStyles'
import { HelperText } from 'components/Forms/HelperText'
import { Icon } from 'components/Icon'
import { Popper } from 'components/Popper'
import { UseClickOutsideElement } from 'hooks/useClickOutsideElement'
import { string, bool, arrayOf, shape, func, oneOf } from 'prop-types'

import {
  OptionsBox,
  HiddenSelect,
  Option,
  OptionIcon,
  CustomSelect,
  IconWrapper,
  Label,
  DisplaySelection
} from './styles'
/**
 * Dropdown é o componente usado para seleção de items dento de uma lista
 */

export const DropdownBasic = ({
  name,
  label,
  value,
  options,
  id,
  status,
  disabled,
  readOnly,
  optionIcon,
  placeholder,
  errorMessage,
  onChange,
  infoMessage,
  optionSize,
  className,
  onBlur
}) => {
  const optionsBoxRef = useRef(null)
  const [active, setActive] = useState(false)
  const [internalValue, setInternalValue] = useState(value)
  const [openOptions, setOpenOptions] = useState(false)
  const [optionIndex, setOptionIndex] = useState(null)
  const selectRef = useRef(null)

  useEffect(() => {
    if (optionIndex === null) {
      optionsBoxRef.current.children[0].focus()
      return
    }

    optionsBoxRef.current.children[optionIndex].focus()
  }, [optionIndex])

  useEffect(() => {
    setInternalValue(value)
  }, [value])

  const handleBlur = useCallback(() => {
    if (active) setActive(false)
    if (openOptions) setOpenOptions(false)

    onBlur?.()
  }, [openOptions])

  const handleClickOption = option => {
    setOpenOptions(false)
    setInternalValue(option.value)
    selectRef.current.focus()

    if (onChange) onChange(option.value)
  }

  const handleOpen = () => {
    setOpenOptions(true)
    if (internalValue) {
      const selectedOptionIndex = options
        .map(option => option.value.indexOf(internalValue))
        .indexOf(0)
      setOptionIndex(selectedOptionIndex)
      setTimeout(() => optionsBoxRef.current.children[selectedOptionIndex].focus(), 100)
    }
  }

  const handleClose = () => {
    setOpenOptions(false)
    selectRef.current.focus()
  }

  const handleButtonClick = () => {
    if (!openOptions) {
      handleOpen()
    } else {
      handleClose()
    }
  }

  const handleArrowChangeState = (prevState, nextState) => {
    if (prevState === null) {
      setInternalValue(options[0].value)
      optionsBoxRef.current.children[0].focus()
      return 0
    }

    if (nextState > options.length - 1 || nextState < 0) {
      setTimeout(() => optionsBoxRef.current.children[prevState].focus(), 100)
      return prevState
    }

    const nextValue = options[nextState].value
    setInternalValue(nextValue)
    setTimeout(() => optionsBoxRef.current.children[nextState].focus(), 100)
    onChange?.(nextValue)
    return nextState
  }

  const keyDownHandler = event => {
    if (!readOnly) {
      const validKeys = ['ArrowUp', 'ArrowDown', 'Enter', 'Escape']
      if (validKeys.indexOf(event.key) !== -1) {
        event.preventDefault()

        if (event.key === 'ArrowDown') {
          if (!openOptions) handleOpen()
          setOptionIndex(prevState => handleArrowChangeState(prevState, prevState + 1))
        }

        if (event.key === 'ArrowUp') {
          if (!openOptions) handleOpen()
          setOptionIndex(prevState => handleArrowChangeState(prevState, prevState - 1))
        }

        if (event.key === 'Enter') {
          if (!openOptions) {
            handleOpen()
          } else handleClickOption(options[optionIndex])
        }
        if (event.key === 'Escape') {
          handleClose()
        }
      }
    }
  }

  return (
    <UseClickOutsideElement handleClick={handleBlur} handleFocusOutside={handleBlur}>
      <BaseWrapper>
        <Popper show={openOptions} useReferenceWidth placement="bottom-start">
          {reference => (
            <CustomSelect
              as="button"
              type="button"
              aria-haspopup="true"
              id={`${id}Label ${id}Content`}
              status={status}
              hasValue={internalValue}
              aria-disabled={disabled}
              aria-readonly={readOnly}
              aria-expanded={openOptions}
              active={active}
              className={className}
              ref={node => {
                reference(node)
                selectRef.current = node
              }}
              tabInde={!(active || openOptions) ? 0 : -1}
              disabled={disabled}
              name={name}
              onFocus={() => setActive(true)}
              onClick={handleButtonClick}
              onKeyDown={event => keyDownHandler(event)}
            >
              <Label
                id={`${id}Label`}
                disabled={disabled}
                status={status}
                active={openOptions}
                hasValue={status === 'default'}
              >
                {label}
              </Label>
              <DisplaySelection
                id={`${id}Content`}
                as="div"
                value={internalValue}
                status={status}
                disabled={disabled}
              >
                {options?.filter(option => option.value === internalValue)[0]?.text || placeholder}
              </DisplaySelection>
              <IconWrapper>
                <Icon
                  icon={openOptions ? 'chevronUp' : 'chevronDown'}
                  styleColor={openOptions ? 'lightBlueCash300' : 'greyCash400'}
                />
              </IconWrapper>
            </CustomSelect>
          )}
          <OptionsBox
            role="list"
            active={openOptions}
            numberOptions={options?.length}
            ref={optionsBoxRef}
          >
            {options?.map(option => (
              <Option
                role="listitem"
                key={option.value}
                onMouseDown={event => event.preventDefault()}
                onClick={() => handleClickOption(option)}
                onKeyDown={event => keyDownHandler(event)}
                selected={internalValue === option.value}
                optionSize={optionSize}
                tabIndex="-1"
              >
                {optionIcon && <OptionIcon icon={optionIcon} size={12} />}
                {option.text}
              </Option>
            ))}
          </OptionsBox>
        </Popper>
        <HiddenSelect id={`select${id}`} name={name} value={internalValue} readOnly>
          {options?.map(option => (
            <option key={option.value} value={option.value}>
              {option.text}
            </option>
          ))}
        </HiddenSelect>
        <HelperText
          status={disabled ? 'default' : status}
          infoMessage={infoMessage}
          errorMessage={errorMessage}
        />
      </BaseWrapper>
    </UseClickOutsideElement>
  )
}

DropdownBasic.propTypes = {
  /**
   * Nome dado ao componente
   */
  name: string.isRequired,
  /**
   * Label dada ao componente
   */
  label: string.isRequired,
  /**
   * Opções recebidas para montar a lista de elementos possivelmente selecionáveis
   */
  options: arrayOf(
    shape({
      value: string.isRequired,
      text: string.isRequired
    })
  ).isRequired,
  /**
   * ID dado ao componente
   */
  id: string.isRequired,
  /**
   * Status do componente
   */
  status: oneOf(['error', 'success', 'default']),
  /**
   * Define se o componente está desabilitado ou não
   */
  disabled: bool,
  /**
   * Ícone das opções do componente
   */
  optionIcon: string,
  /**
   * Nome de placeholder do componente
   */
  placeholder: string,
  /**
   * Mensagem de erro
   */
  errorMessage: string,
  /**
   * Mensagem de informação
   */
  infoMessage: string,
  /**
   * Opção do tamanho das opçoes
   */
  optionSize: oneOf(['large', 'default']),
  onChange: func,
  onBlur: func,
  className: string,
  readOnly: bool,
  value: string
}

DropdownBasic.defaultProps = {
  disabled: false,
  onChange: null,
  onBlur: null,
  status: 'default',
  optionSize: 'default',
  optionIcon: '',
  placeholder: '',
  errorMessage: '',
  infoMessage: '',
  className: '',
  readOnly: false,
  value: ''
}
