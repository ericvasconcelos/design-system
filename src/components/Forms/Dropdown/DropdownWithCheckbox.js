import React, { useState, useRef, useCallback } from 'react'

import { CheckboxBasic } from 'components/Forms/Checkbox'
import { BaseWrapper } from 'components/Forms/commonStyles'
import { HelperText } from 'components/Forms/HelperText'
import { Icon } from 'components/Icon'
import { Popper } from 'components/Popper'
import { UseClickOutsideElement } from 'hooks/useClickOutsideElement'
import { string, bool, arrayOf, shape, func, oneOf } from 'prop-types'

import {
  Button,
  ClearOptions,
  CheckboxOptionsWrapper,
  CheckboxOptionsBox,
  OptionCheckbox,
  CustomSelectCheckbox,
  IconWrapper,
  Label,
  DisplaySelection
} from './styles'
/**
 * DropdownWithCheckbox é o componente usado para seleção de múltiplos itens de uma lista
 */

export const DropdownWithCheckbox = ({
  name,
  label,
  value,
  options,
  id,
  status,
  disabled,
  readOnly,
  placeholder,
  errorMessage,
  onChange,
  infoMessage,
  optionSize,
  className,
  type,
  onBlur,
  clearText
}) => {
  const formatDisplayValue = useCallback(
    valueArray => {
      const numSelected = valueArray.length
      let display = valueArray.map(chosen => chosen.text).join(', ')
      if (numSelected > 2) {
        display = `${numSelected} ${type ? `${type} ` : ''}selecionados(as)`
      }

      return display
    },
    [type]
  )

  const initialDisplayValue = formatDisplayValue(value)
  const [internalValue, setInternalValue] = useState(value)
  const [displayValue, setDisplayValue] = useState(initialDisplayValue)
  const [openOptions, setOpenOptions] = useState(false)
  const selectRef = useRef(null)

  const handleBlur = useCallback(() => {
    if (openOptions) setOpenOptions(false)

    onBlur?.()
  }, [openOptions, onBlur])

  const updateStates = useCallback(
    updatedValue => {
      const newDisplayValue = formatDisplayValue(updatedValue)
      setInternalValue(updatedValue)
      setDisplayValue(newDisplayValue)

      if (onChange) onChange(updatedValue)
    },
    [formatDisplayValue, onChange]
  )

  const addValue = useCallback(
    selected => {
      const defaultOption = options.find(option => option.isDefault)
      const withoutDefault = internalValue.filter(option => option.value !== defaultOption?.value)
      let newInternalValue = [...withoutDefault, selected]

      if (selected.isDefault) newInternalValue = [selected]

      updateStates(newInternalValue)
    },
    [internalValue, options, updateStates]
  )

  const removeValue = useCallback(
    selected => {
      let newInternalValue = internalValue.filter(option => option.value !== selected.value)
      if (newInternalValue.length === 0)
        newInternalValue = options.filter(option => option.isDefault)

      updateStates(newInternalValue)
    },
    [internalValue, options, updateStates]
  )

  const removeAll = useCallback(() => {
    const newInternalValue = options.filter(option => option.isDefault)
    if (internalValue.length > 0) updateStates(newInternalValue)
  }, [internalValue.length, options, updateStates])

  const handleClickOption = useCallback(
    selected => {
      const shouldRemoveSelected = internalValue.some(option => option.value === selected.value)
      if (shouldRemoveSelected) {
        removeValue(selected)
      } else {
        addValue(selected)
      }
    },
    [addValue, internalValue, removeValue]
  )

  const handleOpen = useCallback(() => {
    setOpenOptions(true)
  }, [])

  const handleClose = useCallback(() => {
    setOpenOptions(false)
    selectRef.current.focus()
  }, [selectRef])

  const handleButtonClick = useCallback(() => {
    if (!openOptions) {
      handleOpen()
    } else {
      handleClose()
    }
  }, [handleClose, handleOpen, openOptions])

  const isChecked = useCallback(
    option => internalValue.some(val => val.value === option.value),
    [internalValue]
  )

  return (
    <UseClickOutsideElement handleClick={handleBlur} handleFocusOutside={handleBlur}>
      <BaseWrapper>
        <Popper show={openOptions} useReferenceWidth placement="bottom-start">
          {reference => (
            <CustomSelectCheckbox
              as="button"
              type="button"
              aria-haspopup="true"
              id={`${id}Label ${id}Content`}
              status={status}
              aria-disabled={disabled}
              aria-readonly={readOnly}
              aria-expanded={openOptions}
              className={className}
              ref={node => {
                reference(node)
                selectRef.current = node
              }}
              disabled={disabled}
              name={name}
              onClick={handleButtonClick}
            >
              <Label
                id={`${id}Label`}
                disabled={disabled}
                status={status}
                active={openOptions}
                hasValue={status === 'default'}
              >
                {label}
              </Label>
              <DisplaySelection
                id={`${id}Content`}
                as="div"
                value={displayValue}
                status={status}
                disabled={disabled}
              >
                {internalValue.length ? displayValue : placeholder}
              </DisplaySelection>
              <IconWrapper>
                <Icon
                  icon={openOptions ? 'chevronUp' : 'chevronDown'}
                  styleColor={openOptions ? 'lightBlueCash300' : 'greyCash400'}
                />
              </IconWrapper>
            </CustomSelectCheckbox>
          )}
          <CheckboxOptionsWrapper active={openOptions}>
            <CheckboxOptionsBox numberOptions={options?.length}>
              {options?.map(option => (
                <OptionCheckbox role="listitem" key={option.value} optionSize={optionSize}>
                  <CheckboxBasic
                    id={`${id}checkbox${option.value}`}
                    name={option.text}
                    label={option.text}
                    checked={isChecked(option)}
                    onChange={() => handleClickOption(option)}
                  />
                </OptionCheckbox>
              ))}
            </CheckboxOptionsBox>
            {clearText && (
              <ClearOptions optionSize={optionSize}>
                <Button
                  icon="close"
                  iconSize={16}
                  size="large"
                  type="button"
                  styleType="text"
                  styleColor="blue"
                  onClick={removeAll}
                >
                  {clearText}
                </Button>
              </ClearOptions>
            )}
          </CheckboxOptionsWrapper>
        </Popper>
        <HelperText
          status={disabled ? 'default' : status}
          infoMessage={infoMessage}
          errorMessage={errorMessage}
        />
      </BaseWrapper>
    </UseClickOutsideElement>
  )
}

DropdownWithCheckbox.propTypes = {
  /**
   * Nome dado ao componente
   */
  name: string.isRequired,
  /**
   * Label dada ao componente
   */
  label: string.isRequired,
  /**
   * Opções recebidas para montar a lista de elementos possivelmente selecionáveis
   * Somente um pode ser Default (se todos forem removidos, ele será selecionado).
   */
  options: arrayOf(
    shape({
      value: string.isRequired,
      text: string.isRequired,
      isDefault: bool
    })
  ).isRequired,
  /**
   * ID dado ao componente
   */
  id: string.isRequired,
  /**
   * Status do componente
   */
  status: oneOf(['error', 'success', 'default']),
  /**
   * Define se o componente está desabilitado ou não
   */
  disabled: bool,
  /**
   * Nome de placeholder do componente
   */
  placeholder: string,
  /**
   * Mensagem de erro
   */
  errorMessage: string,
  /**
   * Mensagem de informação
   */
  infoMessage: string,
  /**
   * Opção do tamanho das opçoes
   */
  optionSize: oneOf(['large', 'default']),
  /**
   * Tipo das opções para mensagem de mais de 2 selecionados
   */
  type: string,
  /**
   * Texto que será exibido no botão de limpar o dropdown
   */
  clearText: string,
  onChange: func,
  onBlur: func,
  className: string,
  readOnly: bool,
  value: arrayOf(
    shape({
      value: string.isRequired,
      text: string.isRequired,
      isDefault: bool
    })
  )
}

DropdownWithCheckbox.defaultProps = {
  disabled: false,
  onChange: null,
  onBlur: null,
  status: 'default',
  optionSize: 'default',
  placeholder: '',
  errorMessage: '',
  infoMessage: '',
  className: '',
  type: '',
  clearText: '',
  readOnly: false,
  value: []
}
