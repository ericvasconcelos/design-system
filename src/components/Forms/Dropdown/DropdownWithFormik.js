import React from 'react'

import { useField } from 'formik'
import { string, arrayOf, bool, shape, oneOf, func } from 'prop-types'

import { DropdownBasic } from './DropdownBasic'

/**
 * DropdownWithFormik é o componente usado para seleção de items dento de uma lista.
 * Essa opção já é sendo encapsulada pelo Formik
 */

export const DropdownWithFormik = ({
  name,
  id,
  options,
  label,
  placeholder,
  informationMessage,
  disabled,
  optionSize,
  status,
  className,
  onBlur,
  onChange
}) => {
  const [{ value }, { error }, { setValue, setTouched }] = useField({ name })

  return (
    <DropdownBasic
      name={name}
      id={id}
      value={value}
      errorMessage={error}
      onChange={newValue => {
        setTouched(false)
        setValue(newValue)
        onChange?.(newValue)
      }}
      onBlur={() => {
        setTouched(true)
        onBlur?.()
      }}
      options={options}
      label={label}
      placeholder={placeholder}
      informationMessage={informationMessage}
      status={status}
      disabled={disabled}
      className={className}
      optionSize={optionSize}
    />
  )
}

DropdownWithFormik.propTypes = {
  /**
   * Nome dado ao componente
   */
  name: string,
  /**
   * ID dado ao componente
   */
  id: string,
  /**
   * Label dada ao componente
   */
  label: string,
  /**
   * Nome de placeholder do componente
   */
  placeholder: string,
  /**
   * Mensagem de informação
   */
  informationMessage: string,
  /**
   * Opções recebidas para montar a lista de elementos possivelmente selecionáveis
   */
  options: arrayOf(
    shape({
      value: string.isRequired,
      text: string.isRequired
    })
  ).isRequired,
  /**
   * Define se o componente é válido ou success
   */
  status: oneOf(['error', 'success', 'default']),
  /**
   * Define se o componente está desabilitado ou não
   */
  disabled: bool,
  /**
   * Opção do tamanho das opçoes
   */
  optionSize: oneOf(['large', 'default']),
  className: string,
  onBlur: func,
  onChange: func
}

DropdownWithFormik.defaultProps = {
  name: '',
  label: '',
  id: '',
  placeholder: '',
  informationMessage: '',
  disabled: false,
  optionSize: 'default',
  status: 'default',
  className: '',
  onBlur: null,
  onChange: null
}
