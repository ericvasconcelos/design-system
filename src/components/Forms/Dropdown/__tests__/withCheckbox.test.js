import React from 'react'

import { act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { DropdownWithCheckbox } from 'components/Forms/Dropdown'
import { styledRender } from 'utils/styledRender'

jest.useFakeTimers()
describe('Render DropdownWithCheckbox Component when', () => {
  it('has default props', async () => {
    const { container, getByText, getByLabelText } = styledRender(
      <DropdownWithCheckbox
        id="test"
        name="test"
        label="test1"
        options={[
          { value: 'value1', text: 'text1' },
          { value: 'value2', text: 'text2' }
        ]}
      />
    )

    userEvent.click(getByText('test1'))
    await act(async () => {
      jest.runAllTimers()
    })

    expect(getByLabelText('text1')).not.toBeChecked()
    expect(getByText('text1')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('selects/deselects items', async () => {
    const onChange = jest.fn()
    const { getByText, getByLabelText } = styledRender(
      <DropdownWithCheckbox
        id="test"
        name="test"
        label="test1"
        onChange={onChange}
        options={[
          { value: 'value1', text: 'text1' },
          { value: 'value2', text: 'text2' },
          { value: 'value3', text: 'text3' }
        ]}
      />
    )

    const optionOne = getByLabelText('text1')
    const optionTwo = getByLabelText('text2')
    const optionThree = getByLabelText('text3')

    userEvent.click(getByText('test1'))
    await act(async () => {
      jest.runAllTimers()
    })

    userEvent.click(optionOne)
    userEvent.click(optionTwo)
    userEvent.click(optionOne)

    expect(optionOne).not.toBeChecked()
    expect(optionTwo).toBeChecked()
    expect(optionThree).toBeInTheDocument()
    expect(onChange).toHaveBeenCalledTimes(3)
  })

  it('hides itself with click and blur', async () => {
    const onBlur = jest.fn()
    const { getByText } = styledRender(
      <DropdownWithCheckbox
        id="test"
        name="test"
        label="test1"
        onBlur={onBlur}
        options={[{ value: 'value1', text: 'text1' }]}
      />
    )

    const dropdown = getByText('test1')

    userEvent.click(dropdown)
    userEvent.click(dropdown)
    await act(async () => {
      jest.runAllTimers()
    })

    expect(getByText('text1')).toHaveStyle('visibility: hidden')
    expect(getByText('text1')).not.toBeVisible()

    userEvent.click(dropdown)
    await act(async () => {
      userEvent.click(document.body)
    })

    expect(getByText('text1')).not.toBeVisible()
    expect(onBlur).toHaveBeenCalled()
  })

  it('has isDefault option', async () => {
    const { getByText, getByLabelText } = styledRender(
      <DropdownWithCheckbox
        id="test"
        name="test"
        label="test1"
        options={[
          { value: 'value1', text: 'text1', isDefault: true },
          { value: 'value2', text: 'text2' }
        ]}
      />
    )

    const optionOne = getByLabelText('text1')
    const optionTwo = getByLabelText('text2')

    userEvent.click(getByText('test1'))
    await act(async () => {
      jest.runAllTimers()
    })

    userEvent.click(optionOne)
    userEvent.click(optionTwo)
    userEvent.click(optionTwo)

    await act(async () => {
      jest.runAllTimers()
    })

    expect(optionOne).toBeChecked()
    expect(optionTwo).not.toBeChecked()
  })

  it('selects more than 2', async () => {
    const { getByText, getByLabelText } = styledRender(
      <DropdownWithCheckbox
        id="test"
        name="test"
        label="test1"
        type="opções"
        options={[
          { value: 'value1', text: 'text1' },
          { value: 'value2', text: 'text2' },
          { value: 'value3', text: 'text3' }
        ]}
      />
    )

    const optionOne = getByLabelText('text1')
    const optionTwo = getByLabelText('text2')
    const optionThree = getByLabelText('text3')

    userEvent.click(getByText('test1'))
    await act(async () => {
      jest.runAllTimers()
    })

    userEvent.click(optionOne)
    userEvent.click(optionTwo)
    userEvent.click(optionThree)

    expect(optionOne).toBeChecked()
    expect(optionTwo).toBeChecked()
    expect(optionThree).toBeChecked()
    expect(getByText('3 opções selecionados(as)')).toBeInTheDocument()
  })
})
