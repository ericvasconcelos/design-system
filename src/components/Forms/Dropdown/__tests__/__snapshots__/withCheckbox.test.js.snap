// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render DropdownWithCheckbox Component when has default props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  width: 100%;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  height: 64px;
  gap: 0.5rem;
  position: relative;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: relative;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-sizing: content-box;
  width: calc(100% - 2 * 1rem);
  border-radius: 8px;
  padding: 0.625rem 0.9375rem;
  background-color: #FFFFFF;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  box-shadow: 0px 0px 0px 3px transparent;
  border: 1px solid;
  border-color: #A6B5BB;
}

.c4 {
  color: #003FF2;
  font-size: 0.8125rem;
  font-weight: 500;
  line-height: 1.056rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
  top: unset;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  line-height: 1.0625rem;
  text-align: left;
}

.c6 {
  color: #00224B;
  font-size: 1.0625rem;
  font-weight: 400;
  line-height: 1.381rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
  display: block;
  -webkit-transition: 0.3s;
  transition: 0.3s;
  line-height: 1.375rem;
  text-align: left;
}

.c6:-webkit-autofill,
.c6:-internal-autofill-selected {
  -webkit-animation-name: onAutoFillStart;
  animation-name: onAutoFillStart;
  -webkit-text-fill-color: #455A73;
  box-shadow: inset 0 0 0 100em #FFFFFF;
}

.c6:not(:-webkit-autofill) {
  -webkit-animation-name: onAutoFillCancel;
  animation-name: onAutoFillCancel;
}

.c8 {
  display: block;
  height: 24px;
  width: 24px;
  position: relative;
  margin: 0.5625rem 0px;
}

.c10 {
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/chevron-up.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/chevron-up.svg);
  background-color: #003FF2;
  background-image: none;
  background-size: 24px;
  width: 24px;
  height: 24px;
  -webkit-mask-position-x: 0;
  mask-position-x: 0;
  -webkit-mask-size: 24px;
  mask-size: 24px;
  -webkit-mask-repeat: no-repeat;
  mask-repeat: no-repeat;
}

.c11 {
  visibility: visible;
  pointer-events: all;
  opacity: 1;
  -webkit-transition: visibility 0.1s,opacity ease 0.3s;
  transition: visibility 0.1s,opacity ease 0.3s;
  z-index: 10;
}

.c11[data-Popper-reference-hidden='true'] {
  visibility: hidden;
  pointer-events: none;
}

.c9 {
  grid-area: icon;
}

.c2 {
  display: grid;
  grid-template: 'label icon' 1.0625rem 'input icon' 1.375rem / minmax(0,1fr) 1.5rem;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  row-gap: 0.1875rem;
  position: relative;
  white-space: nowrap;
}

.c3 {
  cursor: pointer;
}

.c5 {
  grid-area: label;
}

.c7 {
  grid-area: input;
  overflow: hidden;
  text-overflow: ellipsis;
  color: #455A73;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #00224B;
  font-size: 1.0625rem;
  font-weight: 400;
  line-height: 1.381rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
  padding: 1rem;
  gap: 0.5rem;
  line-height: 1;
  border-radius: 0.25rem;
  pointer-events: auto;
  height: 2.5rem;
}

.c15 label,
.c15 input {
  cursor: pointer;
}

.c12 {
  display: block;
  background-color: #FFFFFF;
  position: relative;
  height: calc( 2.5rem * undefined + 1rem );
  border-radius: 0.5rem;
  border: 1px solid;
  border-color: #A6B5BB;
}

.c13 {
  height: calc( 2.5rem * 2 + 1rem );
  border-radius: 0.5rem;
  width: 100%;
  box-sizing: border-box;
  z-index: 10;
  overflow-y: auto;
  padding: 0.5rem;
}

.c18 {
  color: #00224B;
  font-size: 0.9375rem;
  font-weight: 400;
  line-height: 1.219rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  font-family: 'DM Sans',sans-serif;
}

.c16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c17 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #FFFFFF;
  border: 2px solid #A6B5BB;
  border-radius: 0.25rem;
  display: inline-block;
  position: relative;
  padding: 10px;
  margin-right: 1rem;
}

.c17[indeterminate] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c17[indeterminate]:after {
  content: '';
  width: 8px;
  height: 2px;
  background-color: #FFFFFF;
  position: absolute;
  top: 9px;
  left: 6px;
  border-radius: 0.25rem;
}

.c17:not([disabled]):hover,
.c17:not([disabled]):focus {
  border-color: #003FF2;
  box-shadow: 0 0 0 4px #003FF23D;
}

.c17[checked] {
  background-color: #01B36A;
  border-color: #01B36A;
}

.c17[checked]:after {
  content: ' ';
  -webkit-mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  mask-image: url(https://qa-static.cashme.com.br/icons/check-bold.svg);
  -webkit-mask-size: contain;
  mask-size: contain;
  background-color: #FFFFFF;
  width: 8px;
  height: 6px;
  position: absolute;
  top: 7px;
  left: 6px;
}

<div>
  <div
    class="c0"
  >
    <button
      aria-disabled="false"
      aria-expanded="true"
      aria-haspopup="true"
      aria-readonly="false"
      class="c1 c2 c3"
      id="testLabel testContent"
      name="test"
      type="button"
    >
      <span
        class="c4 c5"
        id="testLabel"
      >
        test1
      </span>
      <div
        class="c6 c7"
        id="testContent"
        value=""
      />
      <div
        class="c8 c9"
      >
        <div
          aria-hidden="true"
          aria-label=""
          class="c10"
          data-testid="Icon"
        />
      </div>
    </button>
    <div
      class="c11"
      data-popper-escaped="true"
      data-popper-placement="bottom-start"
      data-popper-reference-hidden="true"
      data-testid="Popper"
      style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 10px); width: 0px;"
    >
      <div
        class="c12"
      >
        <ul
          class="c13"
        >
          <li
            class="c14 c15"
            role="listitem"
          >
            <label
              class="c16"
              for="testcheckboxvalue1"
            >
              <input
                class="c17"
                id="testcheckboxvalue1"
                name="text1"
                type="checkbox"
              />
              <span
                class="c18"
                data-testid="Text"
              >
                text1
              </span>
            </label>
            
          </li>
          <li
            class="c14 c15"
            role="listitem"
          >
            <label
              class="c16"
              for="testcheckboxvalue2"
            >
              <input
                class="c17"
                id="testcheckboxvalue2"
                name="text2"
                type="checkbox"
              />
              <span
                class="c18"
                data-testid="Text"
              >
                text2
              </span>
            </label>
            
          </li>
        </ul>
        
      </div>
    </div>
  </div>
</div>
`;
